autonat/autonat-v2.md

```diff
@@ -0,0 +1,193 @@
+# AutonatV2: spec
+
+
+| Lifecycle Stage | Maturity                 | Status | Latest Revision |
+|-----------------|--------------------------|--------|-----------------|
+| 1A              | Working Draft            | Active | r2, 2023-04-15  |
+
+Authors: [@sukunrt]
+
+Interest Group: [@marten-seemann], [@marcopolo], [@mxinden]
+
+[@sukunrt]: https://github.com/sukunrt
+[@marten-seemann]: https://github.com/marten-seemann
+[@mxinden]: https://github.com/mxinden
+[@marcopolo]: https://github.com/marcopolo
+
+
+## Overview
+
+A priori, a node cannot know if it is behind a NAT / firewall or if it is
+publicly reachable. Knowing its NAT status is essential for the node to be
+well-behaved in the network: A node that's behind a NAT / firewall doesn't need
+to advertise its (undialable) addresses to the rest of the network, preventing
+superfluous dials from other peers. Furthermore, it might actively seek to
+improve its connectivity by finding a relay server, which would allow other
+peers to establish a relayed connection.
+
+In `autonat v2` client sends a priority ordered list of addresses. On receiving
+this list the server dials the first address on the list that it is capable of
+dialing. `autonat v2` allows nodes to determine reachability for individual
+addresses. Using `autonat v2` nodes can build an address pipeline where they can
+test individual addresses discovered by different sources like identify, upnp
+mappings, circuit addresses etc for reachability. Having a priority ordered list
+of addresses provides the ability to verify low priority addresses.
+Implementations can generate low priority address guesses and add them to
+requests for high priority addresses as a nice to have. This is especially
+helpful when introducing a new transport. Initially, such a transport will not
+be widely supported in the network. Requests for verifying such addresses can be
+reused to get information about other addresses
+
+Compared to `autonat v1` there are two major differences
+1. `autonat v1` allowed testing reachability for the node. `autonat v2` allows
+testing reachability for an individual address
+2. `autonat v2` provides a mechanism for nodes to verify whether the peer
+actually successfully dialled an address.
+
+
+## AutoNAT V2 Protocol
+
+A node wishing to determine reachability of its adddresses sends a `DialRequest`
+message to a peer on a stream with protocol ID
+`/libp2p/autonat/2.0.0/dial-request`. 
+
+This `DialRequest` message has a list of `AddressDialRequest`s. Each item in
+this list contains an address and a fixed64 nonce. The list is ordered in
+descending order of priority for verfication.
+
+Upon receiving this message the peer attempts to dial the first address from the
```

Comment:
Do I understand correctly that we only ever dial a single address from this list? I think this should be stressed explicitly.

Reply:
Is this line not clear enough
https://github.com/libp2p/specs/blob/62123df073eabbf0a8260be694c36eb59f43b302/autonat/autonat-v2.md?plain=1#L61-L63
?

Reply:
It is clear but at the end of the paragraph. Given the importance of this aspect in the spec, I think it should be very explicit as soon as possible (lookup "frontloading" as a concept).

Reply:
I've moved the must not line near the top of the paragraph and added a clause(https://github.com/libp2p/specs/blob/autonat-v2/autonat/autonat-v2.md?plain=1#L30)
>As the server dials exactly one address from the list, autonat v2 allows nodes to determine reachability for individual addresses. 

in introduction. This should make it clearer.